spec/consensus/gossip/gossip.md

```diff
@@ -0,0 +1,333 @@
+
+In order to ensure convergence even in the presence of failures, the network must be connected in such a way to allow communication around any malicious nodes and to provide redundant paths between correct ones.
+This may not be feasible at all times, but should happen at least during periods in which the system is stable.
+
+In other words, during periods in which messages are timely delivered between correct processes long enough for multiple communication rounds to succeed, processes will identify and solve differences in their states.
+We call "long enough" $\Delta$.
+
+| Eventual $\Delta$-Timely Convergence |
+|---|
+| If $e\in t_p$, for some correct process $p$, at instant $t$, then by $\text{max} \{t,\text{GST}\} + \Delta$, either $e \in t_q$, for every correct process $q$ or $\bar{e} \in t_p$.
+
+$\Delta$ encapsulates the assumption that, during stable periods, timeouts eventually do not expire precociously, given that they all can be adjusted to reasonable values, and the steps needed to converge on entries can be accomplished within $\Delta$.
+Without precocious timeouts, all new entries should help algorithms progress.
+In the Tendermint algorithm, for example, no votes for Nil are added and the round should end if a proposal was made.
+
+
+
+## The tuple space as a CRDT
+
+The two approaches described in the [previous section](#nodes-state-as-a-tuple-space), without the deletion of entries, correspond to operation-based and state-based [Grow-only SET](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#G-Set_(Grow-only_Set)) CRDT (G-Set).
+This distributed data-structure is easily described as a set per process in which elements are added to include them G-Set; the sets kept by processes are approximations of the G-Set.
+
+The [2 Phase-Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#2P-Set_(Two-Phase_Set)) (2P-Set) is a variation that allows removals.
+It combines two sets, one in which elements are added to include them in the 2P-Set, $A$, and one in which elements are added to remove them from the 2P-Set, $D$; the actual membership of the 2P-Set is given by $A \setminus D$.
+
+
+
+
+> Warning/TODO: A word about tombstones, that is $D$
+> - Only state that is not required should be deleted/tombstone'd.
+> - Instead of tombstones, add new entries that trigger removal of other entries (for example, state about a new height); each node must be given information to realize by itself that an entry is no longer needed.
+> - Tombstones are an optimization, kept to prevent data recreation and redeletion.
+> - Tombstones should be garbage collected at some point; imprecision shouldn't affect correctness/termination, since this is an optimization (as long as deleted state is never required again).
+> - Tombstones are not to be gossiped; if they were, they would need to carry proof for the reason they were created, defeating their point.
```

Comment:
2P-Sets require that elements in the "remove" set are communicated between nodes in the same way as elements in the "add" set. If you're not gossiping tombstones (remove set elements) that's fine, but it means you're not implementing a 2P-Set.

Reply:
That is fine. It will eventually be a 2P-Set or I will give it a different name. Remember that these are todo notes on a draft PR.